### DISCUSSION TITLE
Nuovi sviluppi e… un po' di casino

### DISCUSSION THREAD

**Discussion by pdonadeo** [pdonadeo/e-uscito-joypad#44] [Show and tell]
Mi sono messo a modificare pesantemente il progetto per fare quello che ci eravamo detti:

1. utilizzare un database vero e non un enorme file con tutti i dati
2. gestire le chiamate di ricerca come fanno tutti, ovvero con una interfaccia REST che va ad interrogare il DB
3. razionalizzare il frontend, in relazione alle modifiche apportate per il punto 2.

Questo è quindi un messaggio indirizzato principalmente a @MarcoJul: aspetta a fare delle modifiche perché sarebbe impossibile fare merge. Se vuoi un'idea di cosa sta succedendo guarda il [branch PG](https://github.com/pdonadeo/e-uscito-joypad/tree/PG).

Anche far partire il progetto è un po' più complicato: ho introdotto Docker Compose che fa partire:

1. una "ingress" implementata con NGINX
2. un container con PostgreSQL, già inizializzato
3. la app Django del backoffice
4. la app principale

Tutto questo è però molto "work in progress" e on line non ho messo ancora nulla. Soprattutto devo riscrivere il README con le istruzioni per far partire il progetto.

Comunque conto di arrivare ad una situazione "utilizzabile" nel giro di pochi giorni.

**Comment by MarcoJul** [comment]
Ciao Paolo, ok! 

Rimango in attesa di aggiornamenti, nel frattempo buon lavoro!

**Comment by pdonadeo** [comment]
Ho finito il primo giro di modifiche, che riassumo così:
 1. ora sia l'applicazione principale sia il backoffice utilizzano PostgreSQL. È sparito il "mega file JSON";
 2. di conseguenza, per far partire tutto, ho usato Docker Compose che fa partire 4 container:
   a. NGINX
   b. applicazione principale, che serve anche il frontend
   c. PostgreSQL
   d. il backoffice fatto in Django
 3. il frontend è identico nell'aspetto ma non c'è più la chiamata al file JSON che è stata sostituita da un API REST sensato che chiede la lista degli ultimi *n* episodi oppure fa una ricerca per titolo di gioco. Ho quindi tolto Fuse.js, adesso le ricerche si fanno lato server come è giusto che sia.

Operativamente: ho fatto merge di tutto, quindi puoi aggiornati sul mio branch `main` e ripartire da lì.

Il sito on line è quello vecchio ma su [staging](https://staging.euscitojoypad.it/) c'è tutto il software aggiornato.

Siccome ora "È uscito Joypad" ed il suo "backoffice" stanno nella stessa applicazione sono raggiungibili su staging in modo leggermente diverso da prima:
  1. https://staging.euscitojoypad.it/ è il sito (con le novità che ho scritto sopra)
  2. https://staging.euscitojoypad.it/bo/ è il backoffice

**ATTENZIONE** Il backoffice da utilizzare per davvero è ancora quello vecchio a questa URL: https://db.euscitojoypad.it/

https://staging.euscitojoypad.it/bo/ è la stessa cosa ma i dati lì dentro sono per test e quindi potrebbero essere cancellati senza alcuna pietà :joy:

Insomma il data entry del prossimo episodio di Joypad va fatto nel solito vecchio posto: https://db.euscitojoypad.it/

**Reply by MarcoJul** [reply]
Eccomi, sono uscito da un periodo di lavoro intenso (premetto però che la prossima settimana sarò off, ci si rivede lunedì 19) e stasera mi sono rimesso a ricostruire tutta la storia e riaggiornare il mio repository di quanto è successo finora..

Tutto "ok" al momento, a parte che se reinstallo tutto e aggiorno come da branch main di @pdonadeo tutte le chiamate vanno in crash e non trova nessun episodio in listing.. Cerco di capire cosa succede e inizio con le modifiche/aggiornamenti almeno fino a venerdì🖖

**Reply by MarcoJul** [reply]
Per chiarire, io sono al punto attuale del tuo main, @pdonadeo, e ho seguito tutte le procedure precedenti di installazione (con docker e company sul backend e tutto il resto sul frontend)..

**Reply by pdonadeo** [reply]
Rispondo con un po' di ritardo per dirti che devo aggiornare le vecchie istruzioni nel README che ora certamente non funzionano più. Lo faccio questa sera dopo cena.

**Reply by pdonadeo** [reply]
@MarcoJul ho aggiornato il [README](https://github.com/pdonadeo/e-uscito-joypad#readme) con le istruzioni per far partire il progetto in locale sul tuo PC.

**Reply by MarcoJul** [reply]
Ok! Io "purtroppo" sono AFK fino a lunedì prossimo (devo sincronizzare le vacanze con Joypad per l'anno prossimo😅) appena torno mi metto all'opera!

Intanto assegnatemi pure le issue riguardanti il frontend così organizzo le varie richieste!

**Reply by pdonadeo** [reply]
@MarcoJul buone vacanze 😄

**Reply by MarcoJul** [reply]
Ciao! Sono back online, mi sto rimettendo a rimontare il tutto e, oltre a proseguire con il progetto preesistente, a risolvere le varie issue🖖

Vi aggiorno via via (devo ancora sentire l'ultima puntata😭)

**Reply by pdonadeo** [reply]
> Ciao! Sono back online, mi sto rimettendo a rimontare il tutto e, oltre a proseguire con il progetto preesistente, a risolvere le varie issue :vulcan_salute:

Marco fammi un favore: ho riscritto la doc per far partire tutto quanto ma se non ti è chiaro o non funziona qualcosa fammelo sapere subito che cerco di aiutarti.

**Reply by MarcoJul** [reply]
Guarda, stavo giusto cercando di risolvere un problema che si è verificato nella costruzione del backend:

`Error response from daemon: Ports are not available: exposing port TCP 0.0.0.0:5000 -> 0.0.0.0:0: listen tcp 0.0.0.0:5000: bind: address already in use`

Ovviamente a me la porta 5000 risulta libera da ogni altro utilizzo, quindi stavo cercando di capire dove fosse il problema..

**Reply by pdonadeo** [reply]
Quando lanci `docker-compose up --build -d`?

**Reply by MarcoJul** [reply]
Yes, ovviamente `docker compose...` perchè da Mac

Edit: intendo senza trattino

**Reply by MarcoJul** [reply]
<img width="968" alt="image" src="https://user-images.githubusercontent.com/82112168/191612590-ee1edbcd-8af9-4c01-a330-8033644f97f5.png">

Se può aiutare, questa è la situazione dei container attuale, intanto continuo la ricerca del problema

**Reply by pdonadeo** [reply]
Lo sa dio... Domani cerco di corrompere un mio collega col Mac, vediamo se da lui riesco a farlo partire. Purtroppo (o per fortuna...) non ho un Mac da tantissimo tempo quindi non posso provare.

Guggolando a casaccio, prova a vedere se questo ha un senso: [Port 5000 already in use — MacOS Monterey issue](https://medium.com/pythonistas/port-5000-already-in-use-macos-monterey-issue-d86b02edd36c).

**Reply by MarcoJul** [reply]
Ho risolto! In effetti era Airdrop che occupa già di suo quella porta, dovrò fare un po' di switch-on e off in sviluppo😅

So, back online😎

**Reply by pdonadeo** [reply]
Momento: se ti è più comodo cambio porta!

**Reply by MarcoJul** [reply]
Non saprei, finisce che ne becchiamo un'altra occupata da Mac😅

A parte gli scherzi, se non ti crea grossi impicci possiamo provare a indicarne un'altra, anche in ottica di altre persone che fanno il fork via Mac 🤔

**Reply by pdonadeo** [reply]
Ma non scherziamo! Le porte sono tante, eh, scegliamone certamente una comoda per tutti.

Il problema è: che porte libere ci sono su Monterey?

**Reply by MarcoJul** [reply]
Mmh, ci guardo appena riesco e ti dico!

EDIT: che poi ti ho detto che è Airdrop ma è l'Airplay receiver, ma comunque..

**Comment by pdonadeo** [comment]
Altro punto: scrivere un algoritmo di ricerca che risulti davvero efficace è complicato (ma va? altrimenti Larry Page non sarebbe ricco…).

In questi giorni @agofa mi ha dato una mano con dei test sensati.

Agostino: se vuoi provarla trovi la nuova ricerca su staging: https://staging.euscitojoypad.it/

È quella che hai già provato sul mio PC ma con qualche piccola miglioria, almeno si spera 😅

**Reply by agofa** [reply]
Ok
Farò più tardi qualche test

**Reply by agofa** [reply]
@pdonadeo qui un aggiornamento, ho duplicato il tuo coso Gogol e l'ho aggiornato
https://docs.google.com/spreadsheets/d/1f1hwSv9h94rFqb-P-0lj08MtmM536ObCP9aYw5jAaKE/edit?usp=sharing

Vi dico chi è l'assassino: test tutti OK, quelli che ho fatto, ovviamente, direbbe monsieur de La Palice

**Reply by agofa** [reply]
Appunto, trovato forse un problema

**Reply by pdonadeo** [reply]
@agofa Google dice che mi serve l'autorizzazione, guarda la posta.

**Reply by agofa** [reply]
Fatto di proposito, autorizzato

**Comment by pdonadeo** [comment]
Cosa manca? Un po' di documentazione… ma oggi non ho proprio voglia di scrivere un README 😅

**Reply by pdonadeo** [reply]
@agofa Ho modificato ancora la ricerca, secondo me stavolta ci siamo… ma sono pronto ad essere smentito :joy:

**Reply by pdonadeo** [reply]
Dimenticavo: la modifica è già on line sul solito [staging](https://staging.euscitojoypad.it/).

**Reply by agofa** [reply]
Fatto un primo giro, stasera dovrei riuscire a fare test ragionati
Anticipo che ora "legend" trova l'universo conosciuto: 22 giochi
Sto verificando se li sta elencando a caso oppure a ragion veduta

Immagino sia complicato avere una tabella che elenchi i giochi con gli episodi associati
Andrebbe bene anche in forma "pivot", ma non so se mi sono capito

**Reply by pdonadeo** [reply]
@agofa capisco benissimo il problema. Il punto è che in questo sito stiamo facendo una cosa con pochissimo senso: cerchiamo un gioco e mostriamo un elenco di episodi. Il che un senso ce l'ha perché alla fine vogliamo arrivare all'episodio ma è complicato capire con che logica scrivo una cosa e ottengo un elenco di contenuti apparentemente scorrelati da quello che ho scritto.

Per questo ho introdotto due endpoint REST:
 1. [https://staging.euscitojoypad.it/api/search-game-title/STRINGA_DA_CERCARE](https://staging.euscitojoypad.it/api/search-game-title/STRINGA_DA_CERCARE)
    Questo cerca solo i titoli e ti mostra solo il titolo del gioco ed il suo ID nel database. Il problema è che non è molto _user friendly_ perché la risposta è un JSON. Se hai Firefox si riesce a leggere:
    ![image](https://user-images.githubusercontent.com/242049/187469042-b61fd795-6586-4053-8bff-f2dc46a0feb5.png)
    
    Ma se usi Chrome ottieni un cesso:    
    ![image](https://user-images.githubusercontent.com/242049/187469355-cfce786c-73bb-4370-99ab-ba3513207027.png)
    
    Quindi: usa Firefox. I risultati sono in ordine decrescente di rilevanza.
    
 2. Il secondo endpoint permette di ottenere l'elenco degli episodi, dato l'ID del gioco, che ricavi dalla precedente chiamata. Quindi diciamo che hai cercato "legend" ed il secondo gioco che hai in lista è "The Legend of Zelda: Breath of the Wild", ID 36. Adesso fai un'altra chiamata:
    [https://staging.euscitojoypad.it/api/episodes-by-game-id/36](https://staging.euscitojoypad.it/api/episodes-by-game-id/36) (nota il "36" in fondo alla URL)
    
    Il risultato sono gli episodi in cui si è parlato di "The Legend of Zelda: Breath of the Wild" in ordine cronologico dall'episodio più recente:
    ![image](https://user-images.githubusercontent.com/242049/187471124-989307ce-e523-467a-84a0-324714d3fbaf.png)

Con questi due endpoint ti sarà sicuramente più facile fare debug perché cerchi un gioco (primo endpoint) e ottieni un elenco di giochi, in ordine di "rilevanza", che è quello che vuoi.

Poi questi due endpoint io vorrei usarli per cambiare il frontend in modo che un utente inizia a digitare un gioco e si apre la classica tendina di autocompletamento con i nomi dei giochi, non gli episodi. Quando sei soddisfatto clicchi sul nome del gioco nella tendina e a quel punto sappiamo che cavolo di gioco volevi e possiamo selezionare gli episodi col secondo endpoint.

Quest'ultimo paragrafo è un **velato suggerimento** per @MarcoJul 😇

**Reply by agofa** [reply]
Intanto sto finendo la verifica sul famigerato "legend"
Tra poco foglio gogol aggiornato

**Reply by MarcoJul** [reply]
> Quest'ultimo paragrafo è un velato suggerimento per @MarcoJul 😇

Sapevo che si andava a finire su Trenitalia😂 Ma giusto così, l'approccio di questo tipo risolve molti problemi di UX..

**Reply by pdonadeo** [reply]
Pensi di riuscire a farlo? Non so se hai letto come funzionano i due nuovi endpoint.

**Reply by MarcoJul** [reply]
Eh indago e recupero tutti gli argomenti in sospeso, vi tengo aggiornati!
